### PR TITLE
Add more information in case of incompatible constructor signature #2…

### LINKED ISSUE #21253: Need better error message about incompatible constructor signature
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Code**

```ts
class Foo { constructor(a: string) { } }

let FooConstructor: { new(): Foo } = Foo;
```

**Expected behavior:**

Error message should tell the reason why constructor signatures are incompatible.

**Actual behavior:**

```
t.ts(3,5): error TS2322: Type 'typeof Foo' is not assignable to type 'new () => Foo'.
```
This is not particularly helpful, see comments below [this answer](https://stackoverflow.com/a/48312350/43848), especially in comparison to the error message about incompatible methods:

```ts
let foo: { foo(a: string): void };
let foo2: { foo(): void };

foo2 = foo;
```

```
Type '{ foo(a: string): void; }' is not assignable to type '{ foo(): void; }'.
  Types of property 'foo' are incompatible.
    Type '(a: string) => void' is not assignable to type '() => void'.
```

So for constructor signature, ideally, the message could be

```
Type 'typeof Foo' is not assignable to type 'new () => Foo'.
  Types of constructor signature are incompatible.
    Type 'new(a: string) => Foo' is not assignable to type 'new() => Foo'.
```

### PR DESCRIPTION
**Pull request by amaksimovich2**
Fixes #21253 This is my first pull request to TypeScript, so I will free for your comments. At first I tried to implement code from pr of previous contributors, but one test "errorsWithInvokablesInUnions01" failed. So I investigated a problem and added new check on source/target objectFlag comparison. Actually I still not
so good familiar with TS logic, so maybe it's wrong way. If it's true, give me feedback, please.

### CODE REVIEW

**Comment by amaksimovich2** [comment]
@DanielRosenwasser can you or maybe someone else check this pr? Or it will be check as soon as the turn comes?

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2601,6 +2601,10 @@
         "category": "Error",
         "code": 2753
     },
+    "Types of constructor signature are incompatible.": {
+    "category": "Error",
```
Comment (RyanCavanaugh) [inline src/compiler/diagnosticMessages.json]: Please indent the inner properties like the rest of the file

**Comment by RyanCavanaugh** [comment]
@amaksimovich2 overall looks good. Can you resolve the merge conflict and add the repro from #31046 as well? Thanks!

**Comment by sandersn** [comment]
This PR hasn't seen any activity for quite a while, so I'm going to close it to keep the number of open PRs manageable. Feel free to open a fresh PR or continue the discussion here.

**Pull request closed without merging**